src/puppetlabs/kitchensink/file.clj

```diff
@@ -0,0 +1,84 @@
+(ns puppetlabs.kitchensink.file
+  (:import [java.io BufferedWriter FileOutputStream OutputStreamWriter]
+           [java.nio.file CopyOption Files LinkOption Paths StandardCopyOption]
+           [java.nio.file.attribute FileAttribute PosixFilePermissions]))
+
+(defn str->path
+  [path]
+  (Paths/get path (into-array String [])))
+
+(defn get-perms
+  "Returns the currently set permissions of the given file path."
+  [path]
+  (-> (str->path path)
+      (Files/getPosixFilePermissions (into-array LinkOption []))
+      PosixFilePermissions/toString))
+
+(defn set-perms
+  "Set the provided permissions on the given path. The permissions string is in
+  the form of the standard 9 character posix format, ie \"rwxr-xr-x\"."
+  [path permissions]
+  (-> (str->path path)
+      (Files/setPosixFilePermissions
+       (PosixFilePermissions/fromString permissions))
+      (.toFile)))
+
+(defn perms->attribute
+  [permissions]
+  (PosixFilePermissions/asFileAttribute (PosixFilePermissions/fromString permissions)))
+
+(def nofollow-links
+  (into-array LinkOption [LinkOption/NOFOLLOW_LINKS]))
+
+(def ^:private default-permissions
+  (PosixFilePermissions/fromString "rw-r-----"))
+
+(defn atomic-write
+  "Write to a file atomically. This takes a function that should operate on a
+  Writer as its first argument. If permissions are specified (as a string of the
+  form \"rwxrwxrwx\") then the file will be created with those permissions. If
+  not specified and the file already exists then existing permissions will be
+  preserved. If no file exist a default is set."
+  ([path write-function]
+   (atomic-write path write-function nil))
+  ([path write-function permissions]
+   (let [target (str->path path)
+         dir (.getParent target)
+         file-exists? (Files/exists target nofollow-links)
+         owner (if file-exists?
+                 (Files/getOwner target nofollow-links))
+         group (if file-exists?
+                 (Files/getAttribute target "posix:group" nofollow-links))
+         permissions (if permissions
+                       (PosixFilePermissions/fromString permissions)
+                       (if file-exists?
+                         (Files/getPosixFilePermissions target nofollow-links)
+                         default-permissions))
+         attributes (into-array FileAttribute [(perms->attribute "rw-------")])
```

Comment:
I assume this should be 
```suggestion
         attributes (into-array FileAttribute [(perms->attribute permissions)])
```

Reply:
oh I see, you set them below. In that case, I think there should be a comment saying that the real permissions will be set below.